protocol/x/ratelimit/keeper/keeper.go

```diff
@@ -218,6 +221,159 @@ func (k Keeper) SetDenomCapacity(
+		// Since this is in the `EndBlocker`, we log an error instead of panicking.
+		k.Logger(ctx).Error(
+			fmt.Sprintf(
+				"timeSinceLastBlockMilli (%v) <= 0; skipping capacity update. prevBlockTimeMilli = %v, blockTimeMilli = %v",
+				timeSinceLastBlockMilli,
+				prevBlockTimeMilli,
+				blockTimeMilli,
+			),
+		)
+		return
+	}
+
+	// Declare new capacity list to be populated.
+	newCapacityList := make([]dtypes.SerializableInt, len(capacityList))
+
+	for i, limiter := range limitParams.Limiters {
+		// For each limiter, calculate the current baseline.
+		baseline := k.GetBaseline(ctx, supply.Amount.BigInt(), limiter)
+
+		capacityMinusBaseline := new(big.Int).Sub(
+			capacityList[i].BigInt(), // array access is safe because of the invariant check above
+			baseline,
+		)
+
+		// Calculate left operand: `max(baseline, capacity-baseline)`. This equals `baseline` when `capacity <= 2 * baseline`
+		operandL := new(big.Rat).SetInt(
+			lib.BigMax(
+				baseline,
+				capacityMinusBaseline,
+			),
+		)
+
+		// Calculate right operand: `time_since_last_block / period`
+		periodMilli := new(big.Int).Mul(
+			new(big.Int).SetUint64(uint64(limiter.PeriodSec)),
+			big.NewInt(1000),
+		)
+		operandR := new(big.Rat).Quo(
+			new(big.Rat).SetInt(timeSinceLastBlockMilli),
+			new(big.Rat).SetInt(periodMilli),
+		)
+
+		// Calculate: `capacity_diff = max(baseline, capacity-baseline) * (time_since_last_block / period)`
+		// Since both operands > 0, `capacity_diff` is positive or zero (due to rounding).
+		capacityDiff := new(big.Rat).Mul(
+			operandL,
+			operandR,
+		)
+
+		bigRatcapacityMinusBaseline := new(big.Rat).SetInt(capacityMinusBaseline)
+
+		if new(big.Rat).Abs(bigRatcapacityMinusBaseline).Cmp(capacityDiff) < 0 {
+			// if `abs(capacity - baseline) < capacity_diff` then `capacity = baseline``
+			newCapacityList[i] = dtypes.NewIntFromBigInt(baseline)
+		} else if capacityList[i].BigInt().Cmp(baseline) < 0 {
+			// else if `capacity < baseline` then `capacity += capacity_diff`
+			newCapacityList[i] = dtypes.NewIntFromBigInt(
+				new(big.Int).Add(
+					capacityList[i].BigInt(),
+					lib.BigRatRound(capacityDiff, false), // rounds down `capacity_diff`
+				),
+			)
+		} else {
+			// else `capacity -= capacity_diff`
+			newCapacityList[i] = dtypes.NewIntFromBigInt(
+				new(big.Int).Sub(
+					capacityList[i].BigInt(),
+					lib.BigRatRound(capacityDiff, false), // rounds down `capacity_diff`
+				),
+			)
+		}
+
+		// Emit telemetry for the new capacity.
+		telemetry.SetGaugeWithLabels(
+			[]string{types.ModuleName, metrics.Capacity},
+			metrics.GetMetricValueFromBigInt(newCapacityList[i].BigInt()),
+			[]gometrics.Label{
+				metrics.GetLabelForStringValue(metrics.RateLimitDenom, limitParams.Denom),
+				metrics.GetLabelForIntValue(metrics.LimiterIndex, i),
+			},
+		)
+	}
+
+	k.SetDenomCapacity(ctx, types.DenomCapacity{
+		Denom:        limitParams.Denom,
+		CapacityList: newCapacityList,
+	})
```

Comment:
Instead of having a function that:
- Determines the new value
- Emits telemetry
- Writes the new value to state

I suggest the following to be more clear, modular, and unit-testable:
- One function to determine the new value, but not update it in state (call the state update outside of this function)
- Move the telemetry to `SetDenomCapacity` such that it automatically does the telemetry on writes (and you don't have to have telemetry before every time you call `SetDenomCapacity`